// babel-plugin-relay/lib/BabelPluginRelay.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 * @oncall relay
 */

'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var compileGraphQLTag = require('./compileGraphQLTag');
var getValidGraphQLTag = require('./getValidGraphQLTag');
var cosmiconfig = require('cosmiconfig');
var configExplorer = cosmiconfig('relay', {
  searchPlaces: ['relay.config.js', 'relay.config.json', 'package.json'],
  loaders: {
    '.json': cosmiconfig.loadJson,
    '.js': cosmiconfig.loadJs,
    noExt: cosmiconfig.loadYaml
  }
});
var RelayConfig;
var result = configExplorer.searchSync();
if (result) {
  RelayConfig = result.config;
}
/**
 * Using babel-plugin-relay with only the modern runtime?
 *
 *     {
 *       plugins: [
 *         "relay"
 *       ]
 *     }
 */
module.exports = function BabelPluginRelay(context) {
  var t = context.types;
  if (!t) {
    throw new Error('BabelPluginRelay: Expected plugin context to include "types", but got:' + String(context));
  }
  var visitor = {
    TaggedTemplateExpression: function TaggedTemplateExpression(path, state) {
      // Convert graphql`` literals
      var ast = getValidGraphQLTag(path);
      if (ast) {
        compileGraphQLTag(t, path, state, ast);
        return;
      }
    }
  };
  return {
    visitor: {
      Program: function Program(path, state) {
        path.traverse(visitor, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, state), {}, {
          opts: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, RelayConfig), state.opts)
        }));
      }
    }
  };
};
